### PR TITLE
Reduce allocation overhead in quantized and MoE inference

### PR DESCRIPTION
**Pull request by 0xhis**
## Summary

Cut a few temporary allocations from the CPU inference path.

`matmul_qt()` now keeps thread-local scratch space for quantized activations and row scales. The buffers grow when needed and are reused on later calls.

The MoE router now applies sigmoid to the logits in place, so it needs two temporary float arrays instead of three. The expert union bitmap also lives on the stack instead of going through `calloc()` and `free()` once per sparse layer.

The math and routing order are unchanged.

## Validation

- [x] `make -C c check`
- [ ] CUDA changes were tested with `make -C c cuda-test` (not applicable)
- [ ] Performance claims include hardware, commands, and repeatable measurements (no full-model performance claim)

I also ran the repository's tiny GLM oracle:

```sh
SNAP=./glm_tiny REF=./ref_glm.json TF=1 \
  DRAFT=0 RAM_GB=4 OMP_NUM_THREADS=1 \
  ./glm 8 16 16
```

It still matches all teacher-forcing positions:

```text
PREFILL (teacher-forcing) C vs oracolo: 32/32 posizioni
```

A local kernel check showed less overhead for batched int8 calls, while the single-row result was within noise. I left that out as a headline result because it does not represent full-model throughput.

## Compatibility

- [x] The default CPU build remains dependency-free
- [x] No model files, generated binaries, or benchmark artifacts are included

### CODE REVIEW

**Review by JustVugg — APPROVED** [review]
Verified in a worktree: clean build, oracle 32/32 TF + 20/20 greedy (f32), int8 IDOT path bit-identical to main (14/20 vs quantization reference — same as main). The thread-local scratch is the right fix in the right place: matmul_qt is the hottest call site in the engine and this removes a malloc/free per call (we measured allocator/syscall overhead being real in this path when building PILOT). `_Thread_local` keeps it OMP-safe, the overflow guard is correct, and trimming the redundant `sig` buffer in moe() is a nice touch. Approved — merging.
